### PR TITLE
[v7] SequencesAPI refactor

### PR DESCRIPTION
**Pull request by doctrino**
## Description
Please describe the change you have made.

## Checklist:
- [x] Tests added/updated.
- [x] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [x] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite/client/_api/sequences.py [hunk 1]
```diff
@@ -1009,72 +993,108 @@ def delete_range(self, start: int, end: int | None, id: int | None = None, exter
...
 
+    @overload
     def retrieve(
         self,
-        start: int,
-        end: int | None,
-        column_external_ids: SequenceType[str] | None = None,
-        external_id: str | SequenceType[str] | None = None,
-        id: int | SequenceType[int] | None = None,
+        id_or_external_id: int | str,
```
Comment (erlendvollset) [inline cognite/client/_api/sequences.py:1005]: Why this change? It's now inconsistent with the rest of the sdk, as well as other methods in this module.

File: cognite/client/_api/sequences.py [hunk 2]
```diff
@@ -858,7 +839,7 @@ def list(
         return self._list(list_cls=SequenceList, resource_cls=Sequence, method="POST", filter=filter, limit=limit)
 
 
-class SequencesDataAPI(APIClient):
+class SequencesRowsAPI(APIClient):
```
Comment (erlendvollset) [inline cognite/client/_api/sequences.py:842]: Renaming like this makes for a lot  unnecessary noise in the PR. Could it be done in another PR?

File: cognite/client/_api/sequences.py [hunk 3]
```diff
@@ -69,7 +71,8 @@ class SequencesAPI(APIClient):
 
     def __init__(self, config: ClientConfig, api_version: str | None, cognite_client: CogniteClient) -> None:
         super().__init__(config, api_version, cognite_client)
-        self.data = SequencesDataAPI(config, api_version, cognite_client)
+        self.rows = SequencesRowsAPI(config, api_version, cognite_client)
```
Comment (erlendvollset) [inline cognite/client/_api/sequences.py:74]: Why this change? Seems unnecessary. No longer mirrors the time series client, and also devices from the API where it's sequences/data.

**Review by erlendvollset — COMMENTED** [review]
Can you summarize the changes in the PR description and provide rationale for them?

File: cognite/client/_api/sequences.py (hunk 1 above)
Comment (doctrino) [reply cognite/client/_api/sequences.py:1005]: To reflect the API which is either `id` or `external_id`, I know this is no other place in the SDK, but I think `sequences.rows` is a bit unique. The advantage you get is that you get a proper overload implementation which is difficult when you split it into `id` and `external_id`.

File: cognite/client/_api/sequences.py (hunk 2 above)
Comment (doctrino) [reply cognite/client/_api/sequences.py:842]: Sure.

File: cognite/client/_api/sequences.py [hunk 4]
```diff
@@ -1009,72 +993,108 @@ def delete_range(self, start: int, end: int | None, id: int | None = None, exter
...
+        end: int | None = None,
+        columns: SequenceType[str] | None = None,
+        limit: int | None = None,
+    ) -> SequenceRows:
+        ...
+
+    @overload
+    def retrieve(
+        self,
+        id_or_external_id: MutableSequence[int] | MutableSequence[str],
```
Comment (erlendvollset) [inline cognite/client/_api/sequences.py:1016]: MutableSequence is too strict. It excludes types like tuples or other Sequences.

You can use smth like SequenceNotStr from
https://github.com/hauntsaninja/useful_types/blob/main/useful_types/__init__.py if you want to avoid the issue of str also being a Sequence[str].

File: cognite/client/_api/sequences.py (hunk 1 above)
Comment (erlendvollset) [reply cognite/client/_api/sequences.py:1005]: Is there another reason other than getting twice the amount of overloads?

Comment (doctrino) [reply cognite/client/_api/sequences.py:1005]: No, but I struggled to make it work with twice amount of overloads so if you have a solution I am all ears

File: cognite/client/data_classes/sequences.py [hunk 5]
```diff
@@ -33,6 +34,78 @@
...
+        self.last_updated_time = last_updated_time
+
+    @classmethod
+    def load(cls, resource: dict | str, cognite_client: CogniteClient | None = None) -> Self:
+        resource = json.loads(resource) if isinstance(resource, str) else resource
+        # Snake case is supported for backwards compatibility
+        resource = convert_all_keys_to_camel_case(resource)
+        return super().load(resource, cognite_client)
+
+    def as_write(self) -> Self:
```
Comment (erlendvollset) [inline cognite/client/data_classes/sequences.py:78]: I don't love the name `as_write`. Maybe `as_writeable`, although I don't love that either. 

I also think that we should introduce this across the board if we do introduce it, but we can perhaps do it after v7? As far as I can tell it's not breaking.

Comment (doctrino) [reply cognite/client/data_classes/sequences.py:78]: You are right it is not breaking, and it might be good to wait. Will remove it.

File: cognite/client/_api/sequences.py (hunk 4 above)
Comment (doctrino) [reply cognite/client/_api/sequences.py]: Will fix this when #1436 is merged as that has the correct type.

File: cognite/client/_api/sequences.py (hunk 1 above)
Comment (doctrino) [reply cognite/client/_api/sequences.py]: Done one fix.

File: CHANGELOG.md [hunk 6]
```diff
@@ -26,6 +26,20 @@ methods is `json` and `yaml` serializable.
 ### Changed
 * The `CogniteResource._load` has been made public, i.e., it is now `CogniteResource.load`.
 * The `CogniteResourceList._load` has been made public, i.e., it is now `CogniteResourceList.load`.
+* The `client.sequences.data.retrieve` method has changed signature: 
+  The parameter `columns_external_id` is renamed `columns`, the parameters `id` and `external_id` have 
+  been replaced by `id_or_external_id`. This is to better match the API and have a consistent overload implementation.
```
Comment (doctrino) [inline CHANGELOG.md:31]:
```suggestion
  The parameter `columns_external_id` is renamed `columns`, the parameters `id` and `external_id` have 
  been moved to the beginning of the signature. This is to better match the API and have a consistent overload 
  implementation.
```

File: cognite/client/_api/sequences.py (hunk 1 above)
Comment (erlendvollset) [reply cognite/client/_api/sequences.py]: I added a suggestion to how it can be done now.

File: cognite/client/_api/sequences.py (hunk 4 above)
Comment (erlendvollset) [reply cognite/client/_api/sequences.py]: Nice!

File: cognite/client/data_classes/sequences.py [hunk 7]
```diff
@@ -461,36 +615,115 @@ def column_external_ids(self) -> list[str]:
...
-            pandas.DataFrame: The sequence data list as a pandas DataFrame.
+            pandas.DataFrame | dict[str, pandas.DataFrame] | dict[int, pandas.DataFrame]: The sequence data list as a pandas DataFrame.
         """
         pd = local_import("pandas")
-        return pd.concat([seq_data.to_pandas(column_names=column_names) for seq_data in self.data], axis=1)
+        if concat:
+            return pd.concat([seq.to_pandas(column_names) for seq in self], axis=1)
+
+        if key == "external_id":
+            return {seq.external_id: seq.to_pandas(column_names) for seq in self}
```
Comment (haakonvt) [inline cognite/client/data_classes/sequences.py:722]: Should probably have a check here ensuring that all xids exist (if not, all with missing would fight for the same `None` key in the `dict`)

File: cognite/client/utils/_auxiliary.py [hunk 8]
```diff
@@ -56,6 +56,7 @@ def fast_dict_load(
     except TypeError:
         instance = cls()
     # Note: Do not use cast(Hashable, cls) here as this is often called in a hot loop
+    # Accepted: {camel_case(attribute_name): attribute_name}
```
Comment (haakonvt) [inline cognite/client/utils/_auxiliary.py:59]: Test code?
```suggestion
```

**Review by haakonvt — COMMENTED** [review]
Mostly nit + a few comments & suggestions 🤙 Good stuff

File: cognite/client/data_classes/sequences.py [hunk 9]
```diff
@@ -400,42 +548,43 @@ def dump(self, camel_case: bool = False) -> dict[str, Any]:
...
-            values=resource.get("values"),
         )
 
-    def to_pandas(self, column_names: str = "columnExternalId") -> pandas.DataFrame:  # type: ignore[override]
+    def to_pandas(self, column_names: ColumnNames = "columnExternalId") -> pandas.DataFrame:  # type: ignore[override]
         """Convert the sequence data into a pandas DataFrame.
 
         Args:
-            column_names (str):  Which field(s) to use as column header. Can use "externalId", "id", "columnExternalId", "id|columnExternalId" or "externalId|columnExternalId".
+            column_names (ColumnNames): Which field(s) to use as column header. Can use "externalId", "id", "columnExternalId", "id|columnExternalId" or "externalId|columnExternalId".
```
[the remark this answers is not in the document]
Comment (haakonvt) [reply cognite/client/data_classes/sequences.py:577]: (Probably too breaking)

File: CHANGELOG.md [hunk 10]
```diff
@@ -26,6 +26,21 @@ methods is `json` and `yaml` serializable.
...
 * The `CogniteResource._load` has been made public, i.e., it is now `CogniteResource.load`.
 * The `CogniteResourceList._load` has been made public, i.e., it is now `CogniteResourceList.load`.
+* The `client.sequences.data.retrieve` method has changed signature: 
+  The parameter `columns_external_id` is renamed `columns`, the parameters `id` and `external_id` have 
+  been moved to the beginning of the signature. This is to better match the API and have a consistent overload 
+  implementation.
+* The class `SequenceData` has been replaced by `SequenceRows`. The old `SequenceData` class is still available for
+  backwards compatibility, but will be removed in the next major version. However, all API methods now return
+  `SequenceRows` instead of `SequenceData`.
+* The attribute `columns` in `Sequence` has been changed from `SequenceType[dict]` to `SequnceColumnList`.
```
[the remark this answers is not in the document]
Comment (doctrino) [reply CHANGELOG.md]: This has bugged me so much during implementation constantly being confused about SequenceType, this is a much better solution.

File: cognite/client/_api/sequences.py [hunk 11]
```diff
@@ -479,45 +482,22 @@ def create(self, sequence: Sequence | SequenceType[Sequence]) -> Sequence | Sequ
...
-                >>> seq2 = c.sequences.create(Sequence(external_id="my_copied_sequence", columns=seq.columns))
+                >>> seq2 = c.sequences.create(Sequence(external_id="my_copied_sequence", columns=column_def))
 
         """
         assert_type(sequence, "sequences", [SequenceType, Sequence])
-        if isinstance(sequence, SequenceType):
-            sequence = [self._clean_columns(seq) for seq in sequence]
-        else:
-            sequence = self._clean_columns(sequence)
         return self._create_multiple(list_cls=SequenceList, resource_cls=Sequence, items=sequence)
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite/client/_api/sequences.py:499]: Love this suggestion

File: cognite/client/utils/_auxiliary.py (hunk 8 above)
Comment (doctrino) [reply cognite/client/utils/_auxiliary.py:59]: La til kommentar for å forstå hva som skjer fordi det forrvirret meg.

File: cognite/client/_api/sequences.py [hunk 12]
```diff
@@ -1006,75 +990,126 @@ def delete_range(self, start: int, end: int | None, id: int | None = None, exter
...
+            return SequenceRowsList(results)
 
     def retrieve_latest(
         self,
         id: int | None = None,
         external_id: str | None = None,
         column_external_ids: SequenceType[str] | None = None,
         before: int | None = None,
-    ) -> SequenceData:
+    ) -> SequenceRows:
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite/client/_api/sequences.py:1112]: Very good, agreed.

File: cognite/client/data_classes/sequences.py [hunk 13]
```diff
@@ -316,80 +424,120 @@ class SequenceList(CogniteResourceList[Sequence], IdTransformerMixin):
...
+    "id",
+    "columnExternalId",
+    "id|columnExternalId",
+    "externalId|columnExternalId",
+]
+
+_VALID_COLUMN_NAMES = set(get_args(ColumnNames))
+
+
+class SequenceRows(CogniteResource):
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite/client/data_classes/sequences.py:463]: What is the benefit of doing that? I think you still have to overwrite a lot.

File: cognite/client/data_classes/sequences.py [hunk 14]
```diff
@@ -446,12 +595,17 @@ def to_pandas(self, column_names: str = "columnExternalId") -> pandas.DataFrame:
...
+            values.append(row)
+
         return pd.DataFrame(
-            [[x if x is not None else math.nan for x in r] for r in self.values],
-            index=self.row_numbers,
+            values,
+            index=index,
             columns=df_columns,
-        )
+        ).replace({None: pd.NA})
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite/client/data_classes/sequences.py:608]: Of course you did :) Can't help yourself when pandas is involved

File: cognite/client/_api/sequences.py (hunk 1 above)
Comment (doctrino) [reply cognite/client/_api/sequences.py]: Thank you

File: cognite/client/data_classes/sequences.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (haakonvt) [reply cognite/client/data_classes/sequences.py]: 😂 Mostly when I see this pattern:
```py
lst = []
for x in ...:
    lst.append(x)
```
